schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/besr/CodeLine.java

```diff
@@ -0,0 +1,152 @@
+    private final String amount;
+    private final String reference;
+    private final String subscriber;
+
+    /**
+     * All fields will be left padded by zeros if required, if they are complete they will be verified, else a check digit
+     * will be generated
+     *
+     * @param slipType   the type
+     * @param amount     the amount
+     * @param reference  the reference number
+     * @param subscriber the subscriber number
+     */
+    CodeLine(SlipTypeEnum slipType, String amount, String reference, String subscriber) {
+        this.slipType = slipType;
+        this.amount = amount;
+        this.reference = reference;
+        this.subscriber = subscriber;
+    }
+
+    public static boolean isValid(String code) {
+        //TODO: this is actually the reference number
+        if (code == null || code.trim().isEmpty()) {
+            log.trace("Invalid: EMPTY");
+            return false;
+        }
+        String digits = code.replaceAll("[^0-9]", "");
+        if (digits.length() != 27) {
+            log.trace("Invalid: LENGTH={} '{}'", digits.length(), digits);
+            return false;
+        }
+        return CheckDigit.isValid(digits);
+    }
+
+    public static String leftPad(String value, int len) {
+        while (value.length() < len) {
+            value = "0" + value;
+        }
+        return value;
+    }
+
+    public String toCodeLine() throws IllegalCodeLineException {
+        String C = slipType.getCode();
+        String A;
+        if (slipType.isWithAmount()) { // CCAAAAAAAAAAA>RRRRRRRRRRRRRRRRRRRRRRRRRRR+ SSSSSSSSS>
+            if (amount == null) {
+                throw new IllegalCodeLineException("Amount is required but is NULL (slipType=" + slipType + ")");
+            }
+            A = leftPad(onlyDigits(amount), AMOUNT_LENGTH - 1);
+            if (A.length() >= AMOUNT_LENGTH) {
+                throw new IllegalCodeLineException("Amount length exceeds " + (AMOUNT_LENGTH - 1) + " [" + A + "]");
+            }
+            if (A.matches("^0+$")) {
+                throw new IllegalCodeLineException("Amount is required but is ZERO (slipType=" + slipType + ", amount=" + amount + ")");
+            }
+        } else { // CCC>RRRRRRRRRRRRRRRRRRRRRRRRRRR+ SSSSSSSSS>
+            A = "";
+        }
+
+        Preconditions.checkNotNull(reference, "reference");
+        String R = leftPad(onlyDigits(reference), REFERENCE_LENGTH - 1);
+        if (R.length() > REFERENCE_LENGTH) {
+            throw new IllegalCodeLineException("Reference length exceeds " + (REFERENCE_LENGTH - 1) + " [" + R + "]");
+        }
+        if (R.length() == REFERENCE_LENGTH) {
+            if (!CheckDigit.isValid(R)) {
+                throw new IllegalCodeLineException("Reference check digit is invalid [" + R + "]");
+            }
+        } else {
+            R = CheckDigit.complete(R);
+        }
+
+        String S = leftPad(onlyDigits(subscriber), SUBSCRIBER_LENGTH - 1);
```

Comment:
This code could be extracted to a method for both reference and subscriber (String type, int length, String input)